nero-core/src/claim/mod.rs

```diff
@@ -0,0 +1,231 @@
+            claim_challenge_period: ctx.claim_challenge_period,
+            signed_states: ctx.signed_states(),
+        }
+    }
+
+    pub fn challenge_output<C: Verification>(&self, ctx: &Secp256k1<C>) -> TxOut {
+        TxOut {
+            value: DUST_AMOUNT,
+            script_pubkey: Script::new_p2tr(ctx, self.operator_pubkey, None),
+        }
+    }
+
+    pub fn assert_output<C>(&self, ctx: &Secp256k1<C>) -> TxOut
+    where
+        C: Verification,
+    {
+        let taptree = self.taptree(ctx);
+
+        TxOut {
+            value: self.amount,
+            script_pubkey: Script::new_p2tr_tweaked(taptree.output_key()),
+        }
+    }
+
+    fn taptree<C: Verification>(&self, ctx: &Secp256k1<C>) -> TaprootSpendInfo {
+        TaprootBuilder::with_huffman_tree([
+            (9, self.optimistic_payout_script().to_script()),
+            (1, self.assert_script().into_script()),
+        ])
+        .unwrap()
+        .finalize(ctx, *UNSPENDABLE_KEY)
+        .unwrap()
+    }
+
+    pub fn optimistic_payout_script(&self) -> OptimisticPayoutScript {
+        OptimisticPayoutScript::new(self.claim_challenge_period, self.comittee_aggpubkey)
+    }
+
+    pub fn assert_script(
+        &self,
+    ) -> AssertScript<'_, impl Iterator<Item = &'_ SignedIntermediateState>> {
+        AssertScript::new(self.comittee_aggpubkey, self.signed_states.iter())
+    }
+
+    pub fn to_unsigned_tx<C>(&self, ctx: &Secp256k1<C>) -> Transaction
+    where
+        C: Verification,
+    {
+        let challenge_output = self.challenge_output(ctx);
+        let assert_output = self.assert_output(ctx);
+
+        Transaction {
+            // Requires for OP_CSV
+            version: Version::ONE,
+            lock_time: LockTime::ZERO,
+            // Inputs are empty as they should be funded by wallet.
+            input: vec![],
+            output: vec![assert_output, challenge_output],
+        }
+    }
+}
+
+pub struct FundedClaim {
+    /// Inner clain transaction.
+    claim: Claim,
+    /// $x$ - input of the program
+    #[allow(dead_code)]
+    input: Script,
+    /// Funding input got from external wallet.
+    funding_inputs: Vec<TxIn>,
+    /// The change output created after funding the claim tx.
+    change_output: Option<TxOut>,
+}
+
+impl FundedClaim {
+    /// Construct new funded claim transaction.
+    pub fn new(
+        claim: Claim,
+        funding_inputs: Vec<TxIn>,
+        change_output: Option<TxOut>,
+        program_input: Script,
+    ) -> Self {
+        Self {
+            claim,
+            funding_inputs,
+            input: program_input,
+            change_output,
+        }
+    }
+
+    /// Construct bitcoin transaction from funded claim.
+    pub fn to_tx<C>(&self, ctx: &Secp256k1<C>) -> Transaction
+    where
+        C: Verification,
+    {
+        let mut unsigned_tx = self.claim.to_unsigned_tx(ctx);
+
+        // Fullfill the last elements of witness stack
```

Comment:
Are these comments needed?

Reply:
I tried implementing this thing, where we add $x$ at the end of the witness stack:

<img width="659" alt="image" src="https://github.com/user-attachments/assets/6e10ece7-5ebe-430e-97c9-21b0d67ffa92">

But the problem is, that this, of course, doesn't work with Taproot as other stack elements are reserved for different things. Alternatively, we can create a new type of transaction with custom output before **Claim**, that would have P2WSH output with `scriptPubKey`, like this:

```rust
for elem in x {
    OP_DROP
}
OP_CHECKSIG
``` 

But for now, I just decided to comment it out. What do you think?

Reply:
Yeah, fair enough in this case